core/modules/layout/layout.flexible.inc

```diff
@@ -0,0 +1,1118 @@
+      'layout_flexible_template_edit_row_validate',
+    ),
+    '#submit' => array(
+      'layout_flexible_template_edit_row_form_submit',
+    ),
+    '#ajax' => array(
+      'callback' => 'layout_flexible_template_edit_row_ajax',
+    ),
+  );
+  $form['actions']['cancel'] = array(
+    '#type' => 'submit',
+    '#value' => t('Cancel'),
+    '#validate' => array(),
+    '#submit' => array(
+      'layout_flexible_template_cancel_dialog',
+    ),
+    '#ajax' => array(
+      'callback' => 'layout_ajax_flexible_template_cancel_dialog',
+    ),
+  );
+
+  return $form;
+}
+
+/**
+ * Returns the last number of all regions named "Region X" where X is a number.
+ */
+function layout_flexible_template_get_last_region_number($template_data, $form_state = NULL) {
+  $last = 0;
+  if ($form_state) {
+    $region_names = array_filter($form_state['values']['region_names']);
+    $last = $form_state['last_region_number'];
+  }
+  else {
+    foreach ($template_data['rows'] as $row) {
+      foreach ($row['region_names'] as $name) {
+        $region_names[] = $name;
+      }
+    }
+  }
+
+  foreach ($region_names as $region) {
+    $str = explode('Region ', $region);
+    if (!empty($str[1]) && is_numeric($str[1]) && $str[1] > $last) {
+      $last = $str[1];
+    }
+  }
+
+  return $last;
+}
+
+/**
+ * Submit handler for layout_flexible_template_region_style_select().
+ */
+function layout_flexible_template_change_region_style($form, &$form_state) {
+  $form_state['redirect'] = 'admin/structure/layouts/settings/flexible-template/' . $form_state['flexible_template_id'] . '/row/' . $form_state['original_row'] . '/select-widths';
+}
+
+/**
+ * Submit handler for layout_flexible_template_region_style().
+ */
+function layout_flexible_template_region_count_load(&$form, &$form_state) {
+  $form_state['rebuild'] = TRUE;
+}
+
+/**
+ * Submit handler for layout_configure_region_page().
+ */
+function layout_flexible_template_edit_row_validate($form, &$form_state) {
+  $last_region_number = layout_flexible_template_get_last_region_number($form_state['template_data'], $form_state);
+
+  foreach ($form_state['values']['region_names'] as $key => $region_name) {
+    if (empty($region_name)) {
+      $form_state['values']['region_names'][$key] = t('Region') . ' ' . ($last_region_number + $key + 1);
+      $last_region_number++;
+    }
+  }
+}
+
+
+/**
+ * Submit handler for layout_configure_region_page().
+ */
+function layout_flexible_template_edit_row_form_submit($form, &$form_state) {
+  $template_data = $form_state['template_data'];
+  $original_row = $form_state['original_row'];
+
+  $new_row = array(
+    'contains' => $form_state['region_style'],
+    'element' => $form_state['values']['element'],
+    'classes' => $form_state['values']['row_classes'],
+    'container' => $form_state['values']['container'],
+    'region_names' => $form_state['values']['region_names'],
+  );
+
+  if ($original_row == 'add') {
+    $original_row = uniqid();
```

Comment:
Maybe use a UUID instead of `uniqid()`? UUIDs are guaranteed to be unique with no conflicts and it what we use in some other places, e.g. the UUID for placed blocks.

```
    $uuid = new Uuid();
    $original_row = $uuid->generate();
```

Reply:
Uniqids are shorter, and looked better as array keys. But agreed nonetheless. Done.